stdlib/assoc.mc

```diff
@@ -0,0 +1,140 @@
+-- Miking is licensed under the MIT license.
+-- Copyright (C) David Broman. See file LICENSE.txt
+--
+-- A simple library that defines map operations over sequences of tuples.
+
+include "option.mc"
+include "set.mc"
+
+type AssocMap = [(k, v)]
+type AssocTraits = {eq : k -> k -> Bool}
+
+-- 'assocEmpty ()' returns an empty associate map
+let assocEmpty : Unit -> AssocMap = lam _.
+   []
+
+-- 'assocInsert traits k v m' returns a new map, where the key-value pair
+-- ('k','v') is stored. If 'k' is already a key in 'm', its old value will be
+-- overwritten.
+let assocInsert : AssocTraits -> k -> v -> AssocMap -> AssocMap =
+  lam traits. lam k. lam v. lam m.
+    optionMapOrElse (lam _. cons (k,v) m)
+                    (lam i. set m i (k,v))
+                    (index (lam t. traits.eq k t.0) m)
+
+-- 'assocRemove traits k m' returns a new map, where 'k' is not a key. If 'k' is
+-- not a key in 'm', the map remains unchanged after the operation.
+let assocRemove : AssocTraits -> k -> AssocMap -> AssocMap =
+  lam traits. lam k. lam m.
+    optionMapOr m
+                (lam i.
+                  let spl = splitAt m i in
+                  concat spl.0 (tail spl.1))
+                (index (lam t. traits.eq k t.0) m)
+
+
+-- 'assocLookupOpt traits k m' looks up the key 'k' and returns an Option type.
+-- If 'm' has the key 'k' stored, its value is returned, otherwise None () is
+-- returned.
+let assocLookupOpt : AssocTraits -> k -> AssocMap -> OptionV =
+  lam traits. lam k. lam m.
+    optionMapOr (None ())
+                (lam t. Some t.1)
+                (find (lam t. traits.eq k t.0) m)
+
+-- 'assocLookup traits k m' returns the value of key 'k' in 'm'. Throws an error
+-- if 'k' does not exist in 'm'.
+let assocLookup : AssocTraits -> k -> AssocMap -> v =
```

Comment:
I don't remember if we've discussed this at some point in the past, but I would prefer if things that can error have exceptional names, rather than the ones that are total. I'll bring it up at the meeting.

Reply:
Solved during meeting. Option type version is called lookup and the other lookupOrElse.